### PR TITLE
Adding SREM, SSEM and IVIM messages

### PR DESCRIPTION
**Pull request by Irisbus**
I've added SREM and SSEM messages. I think I have it correct, only thing I am not sure with are the files EXTERNAL.c and EXTERNAL.h, my asn1c compiler keeps generating those but I think they may not be needed. Even after removing `/usr/local/share/asn1c`, when I reinstall the asn1c from `riebl/asn1c/integer_uper`, they appear again.

### CODE REVIEW

**Comment by Irisbus** [comment]
I tried adding also IVIM messages. See https://github.com/riebl/vanetza/issues/162. I am manually removing "[[" and "]]" characters from https://standards.iso.org/iso/ts/19091/ed-2/en/ISO-TS-19091-addgrp-C-2018.asn, hope it is okay this way.

File: vanetza/asn1/fetch_iso_asn1_files.cmake [hunk 1]
```diff
@@ -18,4 +18,36 @@ file(READ "${DIRECTORY}/ISO19091.asn" _content)
...
+file(READ "${DIRECTORY}/ISO14906-0-6.asn" _content)
+string(REPLACE "StationType" "StationTypeIso" _content "${_content}")
+file(WRITE "${DIRECTORY}/ISO14906-0-6.asn" "${_content}")
+
+file(DOWNLOAD
+    "https://forge.etsi.org/rep/ITS/asn1/is_ts103301/-/raw/v2.1.1/iso-patched/ISO14823-missing.asn"
+    "${DIRECTORY}/ISO14823.asn")
+
+file(DOWNLOAD
+    "https://forge.etsi.org/rep/LIBS/LibIts/-/raw/798bfce82f199cdfdef4289098bf900832e3932c/asn1/IS/IVI_ISO_TS_17419.asn"
```
Comment (Irisbus) [inline vanetza/asn1/fetch_iso_asn1_files.cmake:50]: Would it be cleaner to use https://standards.iso.org/iso/ts/17419/TS%2017419%20ASN.1%20repository/TS17419_2014_CITSapplMgmtIDs.asn here instead or is it good enough like this? I did not find any newer edition.

Comment (riebl) [reply vanetza/asn1/fetch_iso_asn1_files.cmake:50]: I am fine with that URL. The tricky thing is just to have a compatible set of ASN.1 files…

**Comment by riebl** [comment]
Hi @Irisbus, can you please cross-check my *srem_ssem_ivim* branch? I have rebased your commits on the current master branch and would like to merge them. However, I would feel more confident if someone had a look at it before :-)

**Comment by Irisbus** [comment]
Hi @riebl, unfortunetaly I discovered I cannot encode SSEM messages (both in my pull request and your branch) due to this error:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Can't determine size for unaligned PER encoding of type SSEM because of SignalStatusList sub-type
Aborted
```
When I try to do encode_oer():
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  OER encoding of type SignalStatusMessage failed because of regExtValue sub-type
Aborted
```
I know those errors usually come from not correctly filling some structures, but I believe I have everything filled correctly. Do you have any idea where it might come from? I do not think there are any constraints put on the regional extensions.


Also, I have just a question for you, where does the name ITS_Container_Temperature_t and files ITS-Container_Temperature.{c,h} come from? Do you create them manually (changed from Temperature_t)?

**Comment by Irisbus** [comment]
I am trying to solve that issue as I am not sure if it is not some issue with generated asn code.
Minimum example which is not working:
```
vanetza::asn1::Ssem message;
message->header.messageID = ItsPduHeader__messageID_ssem;
message->header.stationID = 0;
message->header.protocolVersion = 2; // TLC_PROTOCOL_VERSION;

message->ssm.second = 0;

SignalStatus_t* signalStatus = vanetza::asn1::allocate<SignalStatus_t>();
signalStatus->sequenceNumber = 34;
signalStatus->id.id = 0;

SignalStatusPackage_t* packSignalStatus = vanetza::asn1::allocate<SignalStatusPackage_t>();
packSignalStatus->inboundOn.present = IntersectionAccessPoint_PR_lane;
packSignalStatus->inboundOn.choice.lane = 0;
packSignalStatus->status = PrioritizationResponseStatus_unknown;

Reg_SignalStatusPackage_t* regSignalStatusPkg = vanetza::asn1::allocate<Reg_SignalStatusPackage_t>();
regSignalStatusPkg->regionId = 0;
regSignalStatusPkg->regExtValue.present = Reg_SignalStatusPackage__regExtValue_PR_SignalStatusPackage_addGrpC;

packSignalStatus->regional = vanetza::asn1::allocate<SignalStatusPackage::SignalStatusPackage__regional>();
int ret1 = ASN_SEQUENCE_ADD(packSignalStatus->regional, regSignalStatusPkg);

int ret2 = ASN_SEQUENCE_ADD(&signalStatus->sigStatus, packSignalStatus);

Reg_SignalStatusMessage_t* regVal = vanetza::asn1::allocate<Reg_SignalStatusMessage_t>();
regVal->regionId = 0;
regVal->regExtValue.present = Reg_SignalStatusMessage__regExtValue_PR_NOTHING;

message->ssm.regional = vanetza::asn1::allocate<SignalStatusMessage_t::SignalStatusMessage__regional>();
int ret3 = ASN_SEQUENCE_ADD(message->ssm.regional, regVal);

Reg_SignalStatus_t* regSignalStatus = vanetza::asn1::allocate<Reg_SignalStatus_t>();
regSignalStatus->regionId = 0;
regSignalStatus->regExtValue.present =  Reg_SignalStatus__regExtValue_PR_NOTHING;

signalStatus->regional = vanetza::asn1::allocate<SignalStatus::SignalStatus__regional>();
int ret4 = ASN_SEQUENCE_ADD(signalStatus->regional, regSignalStatus);

int ret5 = ASN_SEQUENCE_ADD(&message->ssm.status, signalStatus);
printf("%d, %d, %d, %d, %d\n", ret1, ret2,ret3, ret4, ret5); // Prints 0, 0, 0, 0, 0

std::string error;
if (!message.validate(error))
    std::cerr << "Error: " << error << std::endl;
else
    std::cout << "No error." << std::endl; // Prints no error
//message.encode();   // Fail: Unaligned PER encoding of type SSEM failed because of SignalStatusList sub-type
//vanetza::asn1::encode_per(asn_DEF_SignalStatusMessage, &(message->ssm)); // dtto
//vanetza::asn1::encode_oer(asn_DEF_SignalStatusMessage, &(message->ssm)); // Fail: OER encoding of type SignalStatusMessage failed because of regExtValue sub-type
//vanetza::asn1::encode_per(asn_DEF_SignalStatus, message->ssm.status.list.array[0]); // Fail: OER encoding of type SignalStatusMessage failed because of regExtValue sub-type
```

When I try to encode only `Reg_SignalStatusPackage`:
```
Reg_SignalStatus_t test;
test.regExtValue.present = Reg_SignalStatus__regExtValue_PR_NOTHING;
vanetza::asn1::encode_per(asn_DEF_Reg_SignalStatus, &test); // Fail: Unaligned PER encoding of type Reg-SignalStatus failed because of regExtValue sub-type
```

it fails. However, interestingly, even though this fails:
```
Reg_SignalStatusPackage test;
test.regExtValue.present  = Reg_SignalStatusPackage__regExtValue_PR_NOTHING;
vanetza::asn1::encode_per(asn_DEF_Reg_SignalStatusPackage, &test);
```

this does not:
```
Reg_SignalStatusPackage test;
test.regExtValue.present  = Reg_SignalStatusPackage__regExtValue_PR_SignalStatusPackage_addGrpC;
vanetza::asn1::encode_per(asn_DEF_Reg_SignalStatusPackage, &test);
```

**Comment by Irisbus** [comment]
Another strange thing that I did not notice before is that even though both SREM and SSEM have the regional filed as optional in the ASN definitions, there is not mentioned /* OPTIONAL */ in the regional field in the structures. I do not know if this is not a problem.

**Comment by Irisbus** [comment]
My problem with SSEM messages was caused by me incorrectly filling the regional structures, of course. So that works correctly.

However, I discovered another problem, there is a name collision with IVI messages: there is a type `IssuerIdentifier_t`, which is `long` from ISO14816.asn, but it is a CHOICE structure in IEEE1609.2.asn -- used by Vanetza security code. So I propose renaming the type that is here to something like IVI_IssuerIdentifier, I will try to do that tomorrow.

**Comment by riebl** [comment]
Thanks @Irisbus for your update!
These ASN.1 names tend to clash increasingly as we add more and more ASN.1 sources.

Regarding the *ITS_Container_Temperature_t*: Nothing I have changed explicitly, probably caused by the updated asn1c I have used for code generation.

**Comment by Irisbus** [comment]
I made renaming of `IssuerIdentifier_t` to pull request to your new branch. I am able to encode/decode SREM, SSEM and IVIM messages on Craton/Secton devices now, so from my side this is ready to merge.

**Comment by riebl** [comment]
I have finally merged the SREM, SSEM and IVIM messages into the master branch. Next will be your `IssuerIdentifier_t` PR.

**Pull request closed without merging**